lib/services/service-scaffold.ts

```diff
@@ -0,0 +1,257 @@
+import { WorkerClient } from '../framework/worker-client';
+import { Logger, LogLevel } from '../utils/logger';
+import {
+	ExpressServerDetails,
+	ServiceScaffoldErrorCode, ServiceScaffoldEvent, ServiceScaffoldWorkerEvent
+} from './service-scaffold-types';
+import {
+	ServiceAPIHandle, ServiceEmitContext, ServiceEmitRequest, ServiceEmitResponse,
+	ServiceEmitter, ServiceListener, ServiceRegistration,
+} from './service-types';
+
+/**
+ * A TypedError that encapsulates the subtype of the error and message provided.
+ */
+export class ServiceScaffoldError extends TypedError {
+	/** What category of thing went wrong. */
+	public code: ServiceScaffoldErrorCode;
+
+	/**
+	 * @param code     Sub-type of the error.
+	 * @param message  The error message, is stored in the standard location.
+	 */
+	constructor(code: ServiceScaffoldErrorCode, message: string) {
+		super(message);
+		this.code = code;
+	}
+}
+
+/**
+ * A scaffold class to handle a bunch of the repetitive work associated with being a ServiceListener and ServiceEmitter
+ *
+ * This class provides:
+ * - An express app, if asked for.
+ * - A logger.
+ * - An event registration and handling standard.
+ * - Checks for and extracts data from context.
+ * - A standard way of getting contextualised workers.
+ *
+ * Any inheriting child must:
+ * - Provide a `emitData` method which takes a ServiceEmitContext and returns a promise.
+ * - Provide a `verify` function which should check that incoming data isn't spoofed.
+ * - Provide a `serviceName` getter, which because it is based on file path cannot be inherited away.
+ * - Provide a `apiHandle` getter, which should return the underlying object that executes the requests.
+ * - Enqueue events with `context` and `event`, as per `ServiceScaffoldEvent`.
+ */
+export abstract class ServiceScaffold<T> extends WorkerClient<T> implements ServiceListener, ServiceEmitter {
+	/** Endpoint listener that may be used to, for example, receive web hooks. */
+	protected expressApp?: express.Express;
+
+	/** A place to put output for debug and reference. */
+	private loggerInstance = new Logger();
+
+	/** Store a list of actions to perform when particular actions happen */
+	private eventListeners: { [event: string]: ServiceRegistration[] } = {};
+
+	// https://github.com/resin-io-modules/resin-procbots/issues/262
+	constructor(ingress?: ExpressServerDetails) {
+		super();
+		if (typeof ingress === 'number') {
+			// Construction details provided a port number, so use that.
+			this.expressApp = express();
+			this.expressApp.use(bodyParser.json());
+			this.expressApp.listen(ingress);
+			this.logger.log(LogLevel.INFO, `---> Created Express app on provided port ${ingress} for '${this.serviceName}'.`);
+		} else if (ingress === true) {
+			// Construction details provided a literal true, so default to environment or 443.
+			this.expressApp = express();
+			this.expressApp.use(bodyParser.json());
+			const port = process.env.PORT || 443;
```

Comment:
I'd suggest port `80` instead of `443`. Almost all of our services run as cleartext across the wire and are remapped by the frontends. If nothing else, it makes it *so* much easier to debug when you're developing.